File: pec-observation-app/android-app/app/src/main/kotlin/br/gov/educacao/sp/pecobservation/application/usecases/CreateObservationUseCase.kt
package br.gov.educacao.sp.pecobservation.application.usecases

import br.gov.educacao.sp.pecobservation.domain.entities.Observation
import br.gov.educacao.sp.pecobservation.domain.entities.SyncCommand
import br.gov.educacao.sp.pecobservation.domain.entities.SyncCommandType
import br.gov.educacao.sp.pecobservation.domain.errors.DomainError
import br.gov.educacao.sp.pecobservation.ports.ObservationRepositoryPort
import br.gov.educacao.sp.pecobservation.ports.SyncQueuePort
import java.util.UUID
import javax.inject.Inject

data class CreateObservationInput(
    val schoolId: UUID,
    val teacherId: UUID,
    val subject: String,
    val grade: String,
    val lessonTheme: String,
    val lessonObjectives: String = "",
)

class CreateObservationUseCase @Inject constructor(
    private val repo: ObservationRepositoryPort,
    private val syncQueue: SyncQueuePort,
) {
    suspend fun execute(input: CreateObservationInput): Observation {
        if (input.subject.isBlank())
            throw DomainError.ValidationError("subject", "cannot be empty")
        if (input.grade.isBlank())
            throw DomainError.ValidationError("grade", "cannot be empty")
        if (input.lessonTheme.isBlank())
            throw DomainError.ValidationError("lessonTheme", "cannot be empty")

        val observation = Observation(
            schoolId        = input.schoolId,
            teacherId       = input.teacherId,
            subject         = input.subject,
            grade           = input.grade,
            lessonTheme     = input.lessonTheme,
            lessonObjectives = input.lessonObjectives,
        )

        repo.save(observation)
        syncQueue.enqueue(
            SyncCommand(
                type    = SyncCommandType.CREATE_OBSERVATION,
                payload = mapOf("observationId" to observation.id.toString()),
            )
        )
        return observation
    }
}

File: pec-observation-app/ios-app/Sources/PECObservation/Application/UseCases/CreateObservationUseCase.swift
import Foundation

public struct CreateObservationInput {
    public let schoolId: UUID
    public let teacherId: UUID
    public let subject: String
    public let grade: String
    public let lessonTheme: String
    public let lessonObjectives: String

    public init(schoolId: UUID, teacherId: UUID, subject: String, grade: String,
                lessonTheme: String, lessonObjectives: String = "") {
        self.schoolId = schoolId
        self.teacherId = teacherId
        self.subject = subject
        self.grade = grade
        self.lessonTheme = lessonTheme
        self.lessonObjectives = lessonObjectives
    }
}

public final class CreateObservationUseCase {
    private let repo: ObservationRepositoryPort
    private let syncQueue: SyncQueuePort

    public init(repo: ObservationRepositoryPort, syncQueue: SyncQueuePort) {
        self.repo = repo
        self.syncQueue = syncQueue
    }

    public func execute(input: CreateObservationInput) throws -> Observation {
        guard !input.subject.isEmpty, !input.grade.isEmpty, !input.lessonTheme.isEmpty else {
            throw DomainError.validationFailed("Componente, série e tema da aula são obrigatórios")
        }

        let observation = Observation(
            schoolId: input.schoolId,
            teacherId: input.teacherId,
            subject: input.subject,
            grade: input.grade,
            lessonTheme: input.lessonTheme,
            lessonObjectives: input.lessonObjectives,
            status: .draft
        )
        try repo.save(observation)

        let command = SyncCommand(
            type: .createObservation,
            payload: ["observationId": observation.id.uuidString]
        )
        try syncQueue.enqueue(command)

        return observation
    }
}
